packages/event-handler/src/appsync-graphql/Router.ts

```diff
@@ -0,0 +1,233 @@
+   *   // your business logic here
+   *   return payload;
+   * });
+   
+   * export const handler = async (event, context) =>
+   *   app.resolve(event, context);
+   * ```
+   *
+   * As a decorator:
+   *
+   * @example
+   * ```ts
+   * import { AppSyncGraphQLResolver } from '@aws-lambda-powertools/event-handler/appsync-graphql';
+   *
+   * const app = new AppSyncGraphQLResolver();
+   *
+   * class Lambda {
+   *   @app.onQuery('getPost')
+   *   async handleGetPost(payload) {
+   *     // your business logic here
+   *     return payload;
+   *   }
+   *
+   *   async handler(event, context) {
+   *     return app.resolve(event, context);
+   *   }
+   * }
+   *
+   * const lambda = new Lambda();
+   * export const handler = lambda.handler.bind(lambda);
+   * ```
+   *
+   * @param fieldName - The name of the Query field to register the handler for.
+   * @param handler - The handler function to be called when the event is received.
+   * @param options - Optional route options.
+   * @param options.typeName - The name of the GraphQL type to use for the resolver (defaults to 'Query').
+   */
+  public onQuery(
+    fieldName: string,
+    handler: OnQueryHandler,
+    options?: GraphQlRouteOptions
+  ): void;
+  public onQuery(
+    fieldName: string,
+    options?: GraphQlRouteOptions
+  ): MethodDecorator;
+  public onQuery(
+    fieldName: string,
+    handler?: OnQueryHandler | GraphQlRouteOptions,
+    options?: GraphQlRouteOptions
+  ): MethodDecorator | undefined {
+    if (handler && typeof handler === 'function') {
+      this.onQueryRegistry.register({
+        fieldName,
+        handler,
+        typeName: options?.typeName ?? 'Query',
+      });
+      return;
+    }
+
+    return (_target, _propertyKey, descriptor: PropertyDescriptor) => {
+      const routeOptions = isRecord(handler) ? handler : options;
+      this.onQueryRegistry.register({
+        fieldName,
+        handler: descriptor.value,
+        typeName: routeOptions?.typeName ?? 'Query',
+      });
+      return descriptor;
+    };
+  }
+
+  /**
+   * Register a handler function for the `mutation` event.
+   *
+   * Registers a handler for a specific GraphQL Mutation field. The handler will be invoked when a request is made
+   * for the specified field in the Mutation type.
+   *
+   * @example
+   * ```ts
+   * import { AppSyncGraphQLResolver } from '@aws-lambda-powertools/event-handler/appsync-graphql';
+   *
+   * const app = new AppSyncGraphQLResolver();
+   *
+   * app.onMutation('createPost', async (payload) => {
+   *   // your business logic here
+   *   return payload;
+   * });
+   *
+   * export const handler = async (event, context) =>
+   *   app.resolve(event, context);
+   * ```
+   *
+   * As a decorator:
+   *
+   * @example
+   * ```ts
+   * import { AppSyncGraphQLResolver } from '@aws-lambda-powertools/event-handler/appsync-graphql';
+   *
+   * const app = new AppSyncGraphQLResolver();
+   *
+   * class Lambda {
```

Comment:
Given https://github.com/aws-powertools/powertools-lambda-typescript/issues/3973, do we want to have decorator support in this PR or wait until we have a generic solution to the `this` binding problem that we can use across all these event-handler types.

Reply:
Sounds good.